# model_name = "bert-base-uncased"

**1. Model Class (BertForMaskedLM):**

- This refers to the specific type of BERT model architecture you are using for a particular task.


- **BertForMaskedLM** is a class designed for masked language modeling, where the model tries to predict masked words in a sentence.


- These classes wrap the core BERT architecture and add task-specific layers, like classification heads or masked language modeling heads, so that BERT can be fine-tuned for different purposes.

- Think of these model classes as different versions or extensions of BERT tailored for specific NLP tasks.

- Some of the Known Model Classes

    - BERTforSequenceClassification: Classifiaction

    - BERTforTokenClassifcation  :  NER

    - BERTforQuestionAnswering : Q and A chatbot

    - BERTforMaksedLM : Find the Missing word

**2. Model Name ("bert-base-uncased"):**

- the pre-trained weights and configuration you are using for your model.

- "bert-base-uncased" is one of the many pre-trained models available from Hugging Face.

- It is a base version of BERT with 12 layers, and it is "uncased", meaning it does not distinguish between uppercase and lowercase letters (e.g., "Apple" and "apple" are treated the same).

- There are many different pre-trained model names available, such as "bert-large-uncased", "roberta-base", "gpt-2", etc.

- The model name specifies which set of pre-trained weights and tokenizer to load. This means when you use model_name = "bert-base-uncased", you're downloading a pre-trained model and tokenizer trained on a large corpus of data (like Wikipedia and BooksCorpus).

**Model Class (BertForMaskedLM) defines the task-specific architecture.**

**Model Name ("bert-base-uncased") defines which specific set of pre-trained weights and vocabulary you are loading into that architecture.**

In [ ]:
!pip install transformers

In [ ]:
# import transformers
# dir(transformers)

In [ ]:
from transformers import BertTokenizer, BertForMaskedLM
import torch

#Load pre-trained BERT model and tokenizer
model_name = "bert-base-uncased"
tokenizer = BertTokenizer.from_pretrained(model_name)
model = BertForMaskedLM.from_pretrained(model_name)

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


In [ ]:
sentence="I want to go [MASK]."
tokenized_input = tokenizer.encode_plus(sentence,
                                        return_tensors="pt",
                                        add_special_tokens=True)
tokenized_input
# 101 CLS
# 103 MASK
# 102 SEP
# [CLS], I, want, to, go, [MASK],.,[SEP]
# 1,1,1,1,1,1,1,1
# Model will get internally

{'input_ids': tensor([[ 101, 1045, 2215, 2000, 2175,  103, 1012,  102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1]])}

In [ ]:
sentence1="I want to go [MASK]."
sentence2="I want to go [MASK]."
tokenized_input_two = tokenizer.encode_plus(sentence1,sentence1,
                                        return_tensors="pt",
                                        add_special_tokens=True)
tokenized_input_two

{'input_ids': tensor([[ 101, 1045, 2215, 2000, 2175,  103, 1012,  102, 1045, 2215, 2000, 2175,
          103, 1012,  102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]])}

**token_type_ids**

- will give the indication are we passing single senetnce or multiple sentences

- for single sentences token type ids all are zeros



In [ ]:
sentence="I want to go [MASK]."
tokenized_input_tf = tokenizer.encode_plus(sentence,
                                           return_tensors="tf",
                                           add_special_tokens=True)
tokenized_input_tf

{'input_ids': <tf.Tensor: shape=(1, 8), dtype=int32, numpy=array([[ 101, 1045, 2215, 2000, 2175,  103, 1012,  102]], dtype=int32)>, 'token_type_ids': <tf.Tensor: shape=(1, 8), dtype=int32, numpy=array([[0, 0, 0, 0, 0, 0, 0, 0]], dtype=int32)>, 'attention_mask': <tf.Tensor: shape=(1, 8), dtype=int32, numpy=array([[1, 1, 1, 1, 1, 1, 1, 1]], dtype=int32)>}

In [ ]:
tokenized_input_tf['input_ids'].numpy()

array([[ 101, 1045, 2215, 2000, 2175,  103, 1012,  102]], dtype=int32)

In [ ]:
sentence="I want to go [MASK]."
tokenized_input_np = tokenizer.encode_plus(sentence,
                                           return_tensors="np",
                                           add_special_tokens=True)
tokenized_input_np

{'input_ids': array([[ 101, 1045, 2215, 2000, 2175,  103, 1012,  102]]), 'token_type_ids': array([[0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': array([[1, 1, 1, 1, 1, 1, 1, 1]])}

**Get tokens from token ids**

In [ ]:
tokenized_input['input_ids'][0]  # pytorch tokens

tensor([ 101, 1045, 2215, 2000, 2175,  103, 1012,  102])

In [ ]:
tokenizer.convert_ids_to_tokens(tokenized_input['input_ids'][0])

['[CLS]', 'i', 'want', 'to', 'go', '[MASK]', '.', '[SEP]']

In [ ]:
tokenizer._token_id

100

In [ ]:
# Get the position of the masked token
masked_index = torch.where(tokenized_input["input_ids"] == tokenizer.mask_token_id)[1]
masked_index

tensor([5])

In [ ]:
tokenized_input

{'input_ids': tensor([[ 101, 1045, 2215, 2000, 2175,  103, 1012,  102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1]])}

In [ ]:
def val(**kwargs):
    print(kwargs)

val(name='naresh it',age='25')

{'name': 'naresh it', 'age': '25'}


In [ ]:
def val(**kwargs):
    print(kwargs.items())

val(name='omkar')

dict_items([('name', 'omkar')])


**torch.no_grad**

- we are not using gradient calculations

- gradinet descent is used for updated weights that is for training purpose

- but for predictions again no need to calculation weights

- thats we are using no_grad

In [ ]:
tokenized_input

{'input_ids': tensor([[ 101, 1045, 2215, 2000, 2175,  103, 1012,  102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1]])}

In [ ]:
# It gives probability of each word to become a prediction
with torch.no_grad():
    outputs = model(**tokenized_input)

outputs.logits.shape

torch.Size([1, 8, 30522])

In [ ]:
len(outputs.logits[0])

8

"I want to go [MASK]."

[cls],I,want,to,go,[MASK],'.',[SEP] ===> Model

vocab=30522 words are there
[30522 probabilties]== max prob  ['Cls']
[30522 probabilites]=== max pron should match with  'I'
[30522 probabilites] === max prob word should match with 'want'
[30522 pr] === max prob word:home msk

In [ ]:
len(outputs.logits[0])

8

In [ ]:
import numpy as np
np.argmax(outputs.logits[0][0])

tensor(1012)

In [ ]:
import numpy as np
outputs.logits[0][0]  # 1 word vector
np.argmax(np.array(outputs.logits[0][0]))

np.int64(1012)

In [ ]:
tokenized_input

{'input_ids': tensor([[ 101, 1045, 2215, 2000, 2175,  103, 1012,  102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1]])}

In [ ]:
tokenizer

BertTokenizer(name_or_path='bert-base-uncased', vocab_size=30522, model_max_length=512, is_fast=False, padding_side='right', truncation_side='right', special_tokens={'unk_token': '[UNK]', 'sep_token': '[SEP]', 'pad_token': '[PAD]', 'cls_token': '[CLS]', 'mask_token': '[MASK]'}, clean_up_tokenization_spaces=True, added_tokens_decoder={
	0: AddedToken("[PAD]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	100: AddedToken("[UNK]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	101: AddedToken("[CLS]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	102: AddedToken("[SEP]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	103: AddedToken("[MASK]", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
}
)

In [ ]:
outputs.logits
import pandas as pd
import numpy as np
data=pd.DataFrame(outputs.logits[0])
data.columns=tokenizer.vocab.keys()
data.index=tokenizer.convert_ids_to_tokens(tokenized_input['input_ids'][0])
data

,[PAD],[unused0],[unused1],[unused2],[unused3],[unused4],[unused5],[unused6],[unused7],[unused8],...,##！,##（,##）,##，,##－,##．,##／,##：,##？,##～
[CLS],-6.625476,-6.584863,-6.605818,-6.584106,-6.585435,-6.593922,-6.635152,-6.597618,-6.586466,-6.556528,...,-6.034086,-6.174979,-6.046580,-6.055810,-6.040298,-6.034741,-5.743482,-5.995054,-5.759798,-4.077095
i,-12.754524,-12.413126,-12.508917,-12.276484,-12.244671,-12.947745,-12.369694,-12.405749,-12.336208,-12.501784,...,-11.272758,-10.383535,-10.415506,-10.647338,-12.804343,-10.711677,-10.080683,-9.799524,-9.067732,-12.121728
want,-12.485611,-12.306726,-12.791069,-12.218849,-12.612011,-12.808952,-12.534328,-12.667057,-12.465918,-12.789349,...,-11.393835,-10.397666,-11.566691,-11.027210,-12.691961,-11.607046,-11.043440,-9.087869,-9.851728,-11.984595
to,-16.122074,-15.870565,-16.249203,-15.864260,-15.976259,-16.595772,-15.849690,-16.027554,-16.072828,-15.835457,...,-14.082733,-14.035681,-13.124915,-13.573338,-16.167051,-14.089459,-13.895007,-10.706804,-13.045122,-10.692118
go,-15.401361,-15.331411,-15.362080,-15.463744,-15.471756,-15.820324,-15.210824,-15.645718,-15.107214,-15.489339,...,-13.933654,-13.554896,-13.044665,-13.000900,-14.594225,-13.165329,-12.653934,-11.847961,-12.617706,-12.849973
[MASK],-8.155522,-8.112763,-8.211104,-8.279205,-8.270296,-8.202829,-8.185522,-8.211702,-8.050653,-7.979330,...,-8.278213,-7.376733,-8.202186,-6.454792,-7.996623,-6.576553,-7.401324,-7.180211,-8.109450,-8.045099
.,-12.209925,-11.744507,-12.262812,-12.389926,-12.694874,-12.366922,-12.789671,-12.457346,-12.615365,-12.437675,...,-11.642790,-10.426937,-10.805056,-9.564165,-9.389481,-8.872940,-9.661694,-9.154053,-10.444866,-8.767832
[SEP],-9.463959,-9.342154,-9.401233,-9.367807,-9.361203,-9.482501,-9.502817,-9.390471,-9.408209,-9.301937,...,-8.627596,-8.680964,-8.828867,-8.702608,-8.800778,-8.652872,-8.497360,-8.611466,-8.286394,-6.353649


In [ ]:
# mask
import numpy as np
np.argmax(outputs.logits[0][5])

tensor(2188)

In [ ]:
tokenizer.convert_ids_to_tokens([2188])

['home']

In [ ]:
outputs.logits[0,5]

tensor([-8.1555, -8.1128, -8.2111,  ..., -7.1802, -8.1095, -8.0451])

In [ ]:
# Get the logits for the masked token
predictions = outputs.logits[0, masked_index, :]
predictions

tensor([[-8.1555, -8.1128, -8.2111,  ..., -7.1802, -8.1095, -8.0451]])

In [ ]:
masked_pred=outputs.logits[0,5]
masked_pred

tensor([-8.1555, -8.1128, -8.2111,  ..., -7.1802, -8.1095, -8.0451])

In [ ]:
masked_pred.unsqueeze(0)

tensor([[-8.1555, -8.1128, -8.2111,  ..., -7.1802, -8.1095, -8.0451]])

In [ ]:
import numpy as np
masked_pred=outputs.logits[0,5]
np.argmax(np.array(masked_pred))

np.int64(2188)

In [ ]:
# Get the top predictions
top_indices = torch.topk(masked_pred.unsqueeze(0), 3, dim=1).indices[0].tolist()
top_indices

[2188, 2067, 2185]

In [ ]:
tokenizer.convert_ids_to_tokens([2188,2067,2185])

['home', 'back', 'away']

In [ ]:
from transformers import BertTokenizer, BertForMaskedLM
import torch

#Load pre-trained BERT model and tokenizer
model_name = "bert-base-uncased"
tokenizer = BertTokenizer.from_pretrained(model_name)
model = BertForMaskedLM.from_pretrained(model_name)

# Define a function to predict masked words
def predict_masked_words(sentence):
    # Tokenize the input sentence
    tokenized_input = tokenizer.encode_plus(sentence, return_tensors="pt", add_special_tokens=True)

    # Get the position of the masked token
    masked_index = torch.where(tokenized_input["input_ids"] == tokenizer.mask_token_id)[1]

    # Predict the masked token
    with torch.no_grad():
        outputs = model(**tokenized_input)

    # Get the logits for the masked token
    predictions = outputs.logits[0, masked_index, :]

    # Get the top predictions
    top_indices = torch.topk(predictions, 1, dim=1).indices[0].tolist()

    # Convert token IDs to actual words
    predicted_tokens = [tokenizer.decode([index]) for index in top_indices]

    return predicted_tokens

# Example sentence with a masked word
input_sentence = "I want to go [MASK]."

# Predict masked words
predicted_words = predict_masked_words(input_sentence)

print("Predicted words:", predicted_words)


Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


Predicted words: ['home']


**Using Tensorflow**

In [ ]:
from transformers import BertTokenizer, TFBertForMaskedLM
import tensorflow as tf

# Load pre-trained BERT model and tokenizer
model_name = "bert-base-uncased"
tokenizer = BertTokenizer.from_pretrained(model_name)
model = TFBertForMaskedLM.from_pretrained(model_name)

# Define a function to predict masked words
def predict_masked_words(sentence):
    # Tokenize the input sentence
    tokenized_input = tokenizer.encode_plus(sentence, return_tensors="tf", add_special_tokens=True)

    # Get the position of the masked token
    masked_index = tf.where(tokenized_input["input_ids"] == tokenizer.mask_token_id)

    # Predict the masked token
    outputs = model(tokenized_input)

    # Get the logits for the masked token
    predictions = outputs.logits[0, masked_index[0, 1], :]

    # Get the top predictions
    top_indices = tf.math.top_k(predictions, k=1).indices.numpy()

    # Convert token IDs to actual words
    predicted_tokens = [tokenizer.decode([index]) for index in top_indices]

    return predicted_tokens

# Example sentence with a masked word
input_sentence = "I want to go [MASK]."

# Predict masked words
predicted_words = predict_masked_words(input_sentence)

print("Predicted words:", predicted_words)


/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(
All PyTorch model weights were used when initializing TFBertForMaskedLM.

All the weights of TFBertForMaskedLM were initialized from the PyTorch model.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFBertForMaskedLM for predictions without further training.


Predicted words: ['home']


# model_name = "bert-large-uncased"


In [ ]:
from transformers import BertTokenizer, BertForMaskedLM
import torch

# Load pre-trained BERT model and tokenizer
model_name = "bert-large-uncased"
tokenizer = BertTokenizer.from_pretrained(model_name)
model = BertForMaskedLM.from_pretrained(model_name)

# Define a function to predict masked words
def predict_masked_words(sentence):
    # Tokenize the input sentence
    tokenized_input = tokenizer.encode_plus(sentence, return_tensors="pt", add_special_tokens=True)

    # Get the position of the masked token
    masked_index = torch.where(tokenized_input["input_ids"] == tokenizer.mask_token_id)[1]

    # Predict the masked token
    with torch.no_grad():
        outputs = model(**tokenized_input)

    # Get the logits for the masked token
    predictions = outputs.logits[0, masked_index, :]

    # Get the top 5 predictions
    top_5_indices = torch.topk(predictions, 5, dim=1).indices[0].tolist()

    # Convert token IDs to actual words
    predicted_tokens = [tokenizer.decode([index]) for index in top_5_indices]

    return predicted_tokens

# Example sentence with a masked word
input_sentence = "Are you going to [MASK]."
# Predict masked words
predicted_words = predict_masked_words(input_sentence)

print("Predicted words:", predicted_words)


tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/571 [00:00<?, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


model.safetensors:   0%|          | 0.00/1.34G [00:00<?, ?B/s]

Some weights of the model checkpoint at bert-large-uncased were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


Predicted words: ['?', 'die', 'sleep', '...', 'stay']


# model_name = bert-base-multilingual-uncased

In [ ]:
from transformers import BertTokenizer, BertForMaskedLM
import torch

# Load pre-trained BERT model and tokenizer
model_name = "bert-base-multilingual-uncased"
tokenizer = BertTokenizer.from_pretrained(model_name)
model = BertForMaskedLM.from_pretrained(model_name)

# Define a function to predict masked words
def predict_masked_words(sentence):
    # Tokenize the input sentence
    tokenized_input = tokenizer.encode_plus(sentence, return_tensors="pt", add_special_tokens=True)

    # Get the position of the masked token
    masked_index = torch.where(tokenized_input["input_ids"] == tokenizer.mask_token_id)[1]

    # Predict the masked token
    with torch.no_grad():
        outputs = model(**tokenized_input)

    # Get the logits for the masked token
    predictions = outputs.logits[0, masked_index, :]

    # Get the top 5 predictions
    top_5_indices = torch.topk(predictions, 5, dim=1).indices[0].tolist()

    # Convert token IDs to actual words
    predicted_tokens = [tokenizer.decode([index]) for index in top_5_indices]

    return predicted_tokens

# Example sentence with a masked word
input_sentence = "What is your favorite [MASK]?"
# Predict masked words
predicted_words = predict_masked_words(input_sentence)

print("Predicted words:", predicted_words)


Some weights of the model checkpoint at bert-base-multilingual-uncased were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


Predicted words: ['game', 'food', 'thing', 'song', 'movie']
